api/rest/coordinator.go

```diff
@@ -0,0 +1,231 @@
+package rest
+
+import (
+	"archive/zip"
+	"context"
+	"fmt"
+	"io"
+	"io/ioutil"
+	"time"
+
+	"github.com/sirupsen/logrus"
+)
+
+const numberOfWorkers = 10
+
+type BundleStatus struct {
+	ID   string
+	node node
+	done bool
+	err  error
+}
+
+// Coordinator is an interface to coordinate the creation of diagnostics bundles
+// across a cluster of nodes
+type Coordinator interface {
+	Create(ctx context.Context, id string, nodes []node) <-chan BundleStatus
+	Collect(ctx context.Context, bundleID string, numBundles int, statuses <-chan BundleStatus) (string, error)
+}
+
+// ParallelCoordinator implements Coordinator interface to coordinate bundle
+// creation across a cluster, parallelized across multiple goroutines
+type ParallelCoordinator struct {
+	client Client
+}
+
+// job is a function that will be called by worker. The output will be added to results chanel
+type job func() BundleStatus
+
+// worker is a function that will run incoming jobs from jobs channel and put jobs output to results chan
+func worker(ctx context.Context, jobs <-chan job, results chan<- BundleStatus) {
+	for {
+		select {
+		case <-ctx.Done():
+			break
+		case j := <-jobs:
+			results <- j()
+		}
+	}
+}
+
+// NewParallelCoordinator constructs a new parallelcoordinator.
+func NewParallelCoordinator(client Client) ParallelCoordinator {
+	return ParallelCoordinator{
+		client: client,
+	}
+}
+
+// Create starts bundle creation process. Creation process could be monitor on returned channel.
+func (c ParallelCoordinator) Create(ctx context.Context, id string, nodes []node) <-chan BundleStatus {
+
+	jobs := make(chan job)
+	statuses := make(chan BundleStatus)
+
+	for i := 0; i < numberOfWorkers; i++ {
+		go worker(ctx, jobs, statuses)
+	}
+
+	for _, n := range nodes {
+		logrus.WithField("IP", n.IP).Info("Sending creation request to node.")
+
+		// necessary to prevent the closure from giving the same node to all the calls
+		tmpNode := n
+		jobs <- func() BundleStatus {
+			//TODO(janisz): Handle context
+
+			// because this will also make a request from the coordinating master, we
+			// can't tell it to make a local bundle with the same ID this guarantees that
+			// the coordinating master's local bundle has a different ID.
+			// from this point, the full <ip>-<id> will be carried in the BundleStatus ID field so this does not
+			// need to be recalculated in the coordinator
+			fullID := fmt.Sprintf("%s-%s", tmpNode.IP, id)
```

Comment:
Why not using the same ID for all local bundles?

Reply:
We have a few possibilities here but the main thing is to avoid using the same ID as the remote bundle, this was just the quickest to write.

Reply:
I still don't get why we don't use UUIDs here as I asked in https://github.com/dcos/dcos-diagnostics/pull/117#discussion_r290715769 and as the design doc says. So either we change the design doc or just use UUIDs.